### PR TITLE
Add zip_subtrees for paired iteration over DataTrees

### PR DESCRIPTION
**Pull request by shoyer**
This should be used for implementing DataTree arithmetic inside map_over_datasets, so the result does not depend on the order in which child nodes are defined.

I have also added a minimal implementation of breadth-first-search with an explicit queue the current recursion based solution in xarray.core.iterators (which has been removed). The new implementation is also slightly faster in my microbenchmark:

    In [1]: import xarray as xr

    In [2]: tree = xr.DataTree.from_dict({f"/x{i}": None for i in range(100)})

    In [3]: %timeit _ = list(tree.subtree)
    # on main
    87.2 μs ± 394 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

    # with this branch
    55.1 μs ± 294 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

- [x] Tests added

### CODE REVIEW

**Comment by shoyer** [comment]
I made a pass at re-implementing `map_over_datasets` using `zip_subtree` in https://github.com/shoyer/xarray/pull/2

File: xarray/tests/test_treenode.py [hunk 1]
```diff
@@ -299,15 +303,12 @@ def create_test_tree() -> tuple[NamedNode, NamedNode]:
     return a, f
 
 
-class TestIterators:
+class TestZipSubtrees:
 
-    def test_levelorderiter(self):
+    def test_one_tree(self):
```
Comment (headtr1ck) [inline xarray/tests/test_treenode.py:308]:
```suggestion
    def test_one_tree(self) -> None:
```

**Review by headtr1ck — APPROVED** [review]
Only some minor typing remarks, the rest looks good!

File: xarray/core/treenode.py [hunk 2]
```diff
@@ -773,3 +777,41 @@ def _path_to_ancestor(self, ancestor: NamedNode) -> NodePath:
...
+    queue = collections.deque([trees])
+
+    while queue:
+        active_nodes = queue.popleft()
+
+        # yield before raising an error, in case the caller chooses to exit
+        # iteration early
+        yield active_nodes
+
+        first_node = active_nodes[0]
```
[the remark this answers is not in the document]
Comment (shoyer) [reply xarray/core/treenode.py:804]: Good point, I added an error to catch this case.

File: xarray/tests/test_treenode.py [hunk 3]
```diff
@@ -317,8 +318,37 @@ def test_levelorderiter(self):
             "g",
             "i",
         ]
+        result = [node[0].name for node in zip_subtrees(root)]
         assert result == expected
 
+    def test_different_order(self):
+        first: NamedNode = NamedNode(
```
[the remark this answers is not in the document]
Comment (shoyer) [reply xarray/tests/test_treenode.py:325]: mypy does seem to struggle here -- it raises an error about missing type annotations.

I have not been able to precisely reproduce the mypy setup from CI on my local machine, so I'm going to save this existing issue for someone else to look into.

[the remark this answers is not in the document]
Comment (headtr1ck) [reply xarray/tests/test_treenode.py:325]: I can only see the issue of not passing the generic type to the parent class like I said before. But ofc, we can keep this open by now.

File: xarray/core/treenode.py [hunk 4]
```diff
@@ -395,15 +396,18 @@ def subtree(self: Tree) -> Iterator[Tree]:
...
         """
-        from xarray.core.iterators import LevelOrderIter
-
-        return LevelOrderIter(self)
+        # https://en.wikipedia.org/wiki/Breadth-first_search#Pseudocode
+        queue = collections.deque([self])
+        while queue:
+            node = queue.popleft()
+            yield node
+            queue.extend(node.children.values())
```
Comment (TomNicholas) [inline xarray/core/treenode.py:410]: Replacing the entire `iterators.py` file with 6 lines is so clever it's almost rude 🤣

**Review by TomNicholas — COMMENTED** [review]
> so the result does not depend on the order in which child nodes are defined.

I'm not sure I understand this. Surely `zip_subtrees` is zipping nodes according to the order they appear in `.children`, which is the order they are defined?

**Comment by shoyer** [comment]
> > so the result does not depend on the order in which child nodes are defined.
> 
> 
> I'm not sure I understand this. Surely `zip_subtrees` is zipping nodes according to the order they appear in `.children`, which is the order they are defined?

Let me restate this: zip_subtrees allows for zipping together multiple DataTree objects even if child nodes on different trees are defined in different orders, as long as the sets of each node's children match.

**Comment by TomNicholas** [comment]
And when you say "match" you mean the set of names of the children on tree A match the set of names of the children on tree B?

**Comment by shoyer** [comment]
> And when you say "match" you mean the set of names of the children on tree A match the set of names of the children on tree B?

Exactly, matching is based on relative path from each root.

**Comment by TomNicholas** [comment]
> matching is based on relative path from each root.

So I think this will basically create a breaking change relative to how `map_over_subtree` (and hence all arithmetic) used to work in `xarray-contrib/datatree`. In the old model the definition of isomorphic was such that you could actually multiply two trees with the same structure but differently-named nodes, e.g.

```python
dt1 = DataTree.from_dict({'a': ..., 'a/b': ..., 'a/c': ...})
dt2 = DataTree.from_dict({'e': ..., 'e/f': ..., 'e/g': ...})

dt1 * dt2  # would return a tree with names following dt1
```

I was never really sure if that generality was actually necessary though.

Your new definition of `isomorphic` is sort of like having `strict_names=True` always, except that the as long as the set of names match then their names will be used to determine their corresponding nodes in the other tree, regardless of order.

Therefore if we're relaxing this then I think there is no longer any need to think of the data model of datatree as being an _ordered_ set of children, as no mapping behaviour will depend on that order any longer.

I think it's okay to change this behaviour, including getting rid of the `strict_names` kwarg. After all I have never once needed the generality above! I just wanted to really spell out the implications to make sure I and anyone reading is following.

The last sentence of [this section](https://github.com/pydata/xarray/blob/main/doc/user-guide/hierarchical-data.rst#comparing-trees-for-isomorphism) of the docs will need changing, and the change would merit an entry in the migration guide.

**Review by TomNicholas — APPROVED** [review]
Now that I think I understand the behaviour change here, the implementation is great! I really like the idea of `zip_subtrees` as a primitive.

**Comment by shoyer** [comment]
> The last sentence of [this section](https://github.com/pydata/xarray/blob/main/doc/user-guide/hierarchical-data.rst#comparing-trees-for-isomorphism) of the docs will need changing, and the change would merit an entry in the migration guide.

I'll do this in the next PR when I migrate over the `map_over_datasets` stuff.

**Pull request merged**